### PR TITLE
Add pruning functionality to tokenizer creation to remove zero frequency tokens

### PR DESCRIPTION
**Pull request by chandanms**
## Description
This PR implements tokenizer pruning functionality to address high cosine similarity issues between token embeddings. The solution removes unused tokens from the trained tokenizer vocabulary and reassigns sequential token IDs to create a more compact and efficient tokenizer.

- Added prune_tokenizer() function that identifies and removes tokens with zero frequency in the dataset
- Preserves special tokens ([UNK], [EOS]) and their functionality, original tokenizer settings (normalizers, pre-tokenizers, post-processors, decoders) and serialized IDs

Other minor updates in the PR are 

- changing the default names to more updated ones (ex: `lennart-finke/SimpleStories` to `SimpleStories/SimpleStories`)
- streamlining the tokenizer creating steps like removing the dataset splitting (it adds no value in case of tokenizer), making the script easily executable with different datasets etc

## Related Issue
Fixes the token embedding cosine similarity issue where ~2% of zero-frequency tokens from WordPieceTrainer were collapsing into the same embedding direction as [UNK], causing cosine similarities >0.9 and nearly identical embeddings.

## Motivation and Context
The original issue was that trained model's embedding layer had cosine similarities >0.9. Investigation revealed that WordPieceTrainer was generating tokens with zero frequency in the actual dataset, which during model training would collapse to have similar embeddings as the [UNK] token. This resulted in high cosine similarities between token embeddings.

The pruning approach removes these problematic tokens and creates a cleaner, more efficient tokenizer that only contains tokens actually present in the training data.

## How Has This Been Tested?
A test code has been written in `test_tokenizer.py` which demonstrates the wordpiece tokenizer issue of creating artifacts when vocab_size is too high and how pruning removes the zero frequency tokens

## Does this PR introduce a breaking change?
No. The pruning step can be optionally disabled by not calling prune_tokenizer() in the main workflow.

### CODE REVIEW

File: simple_stories_train/35M_config.yaml [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
   is_tokenized: false
-  tokenizer_file_path: tokenizer/simplestories-4096.json
+  tokenizer_file_path: tokenizer/simplestories-tokenizer.json
   split: test
   streaming: false
   n_ctx: 512
   seed: 0
   column_name: story
-model_name: 35M
+model_name: 1.25M
```
Comment (danbraunai) [inline simple_stories_train/35M_config.yaml:19]: Not great having the file be called 35M but the model name 1.25M.

File: simple_stories_train/tokenizer.py [hunk 2]
```diff
@@ -23,20 +25,33 @@
 COMMON_SUFFIXES = ["ed", "ing", "ly", "er", "ness"]
 
 
-def clean_dataset(dataset="lennart-finke/SimpleStories") -> DatasetDict:
+def clean_dataset(dataset="SimpleStories/SimpleStories") -> list[str]:
```
Comment (danbraunai) [inline simple_stories_train/tokenizer.py]: I don't think dataset should have a default here. Can remove.

File: simple_stories_train/tokenizer.py [hunk 3]
```diff
@@ -23,20 +25,33 @@
...
     dataset = load_dataset(dataset, trust_remote_code=False)
     trans = str.maketrans(
         {"\u201d": '"', "\u201c": '"', "\u2019": "'", "\u2018": "'", "\u2014": "-", "\u2026": "..."}
     )
 
-    cleaned = [
+    cleaned_data = [
         s.translate(trans).encode("ascii", "ignore").decode("ascii").lower()
         for s in dataset["train"]["story"]  # pyright: ignore
     ]
```
Comment (danbraunai) [inline simple_stories_train/tokenizer.py]: I know this wasn't done in this PR, but this code puts the entire dataset in memory. It would be better to clean this in an iterator to avoid this.

File: simple_stories_train/tokenizer.py [hunk 4]
```diff
@@ -89,37 +103,112 @@ def create_tokenizer(vocab_size=4096) -> Tokenizer:
...
 
-def test_tokenizer(filepath: str, dataset: str = "lennart-finke/SimpleStories") -> None:
+    Returns:
+        Pruned Tokenizer object with sequential IDs
+    """
+    original_vocab_size = len(tokenizer.get_vocab())
+    print(f"Original vocabulary size: {original_vocab_size}")
+
+    # Always keep special tokens (IDs 0 and 1)
+    special_tokens = {0, 1}
```
Comment (danbraunai) [inline simple_stories_train/tokenizer.py]: A general tokenizer might not have exactly 0 and 1 as special tokens. Perhaps there's a method on the Tokenizer class to get the special tokens?

File: simple_stories_train/tokenizer.py [hunk 5]
```diff
@@ -89,37 +103,112 @@ def create_tokenizer(vocab_size=4096) -> Tokenizer:
...
+
+    print(f"New vocabulary size: {len(new_vocab)}")
+
+    # Create new tokenizer
+    new_tokenizer: Tokenizer = Tokenizer(WordPiece(vocab=new_vocab, unk_token="[UNK]"))  # type: ignore
+
+    # Add special tokens back
+    new_tokenizer.add_special_tokens(
+        [AddedToken("[UNK]", special=True), AddedToken("[EOS]", special=True)]
+    )
```
Comment (danbraunai) [inline simple_stories_train/tokenizer.py]: Seems a bit weird that you specify the [UNK] token in both of these calls. Are they both necessary?

Comment (danbraunai) [reply simple_stories_train/tokenizer.py]: In general, tokenizer issues can hide away for a long time and can be very costly, so would be good to make sure that stuff like this is done correctly and do several sanity checks and concrete tests.

File: simple_stories_train/tokenizer.py [hunk 6]
```diff
@@ -169,15 +264,25 @@ def print_split_words(story_tokens: list[str]) -> None:
             print(f"{word_start} {' '.join(word_parts)}")
 
 
-def analysis(vocab_size=4096) -> None:
-    tokenizer = load_tokenizer(vocab_size)
-    stories = clean_dataset(dataset="lennart-finke/SimpleStories")["validation"]["story"]
-    tokenized_stories = [tokenizer.encode(story).tokens for story in stories]
+def analysis(tokenizer_name="simplestories-4096.json") -> None:
```
Comment (danbraunai) [inline simple_stories_train/tokenizer.py]: This function isn't called anywhere. Can remove or put in a test or PR description or something.

**Review by danbraunai — APPROVED** [review]
Nice, feel free to merge after addressing my comments.

I commented on this once, but I think all default arguments should be removed that aren't actually used in this codebase. I spotted a few of them.

I changed the base of this PR to a dev branch. I don't think we should change the main branch, at least not right now, as people will expect that to match the paper.

File: simple_stories_train/tokenizer.py (hunk 5 above)
Comment (chandanms) [reply simple_stories_train/tokenizer.py]: I rechecked the docs and you are right. For pruning, there is no need to do both. I have put a comment on the revised version of the script to clarify this.

File: simple_stories_train/tokenizer.py (hunk 4 above)
Comment (chandanms) [reply simple_stories_train/tokenizer.py]: There is no direct method but I have added a function to do this now

File: simple_stories_train/tokenizer.py (hunk 3 above)
Comment (chandanms) [reply simple_stories_train/tokenizer.py]: Good point. Made it an iterator but now its being called twice: once to train and once to prune which I think is okay. The tradeoff between time and memory is not big.

I spotted that this only returned "train" split of the data which I think is not right. So I changed it to return all splits of the data and that way the tokenizer can train on the complete dataset.

File: simple_stories_train/35M_config.yaml (hunk 1 above)
Comment (chandanms) [reply simple_stories_train/35M_config.yaml:19]: I think its better to change the filename to train_config.yaml instead

Comment (danbraunai) [reply simple_stories_train/35M_config.yaml:19]: Yep that sounds fine. Do you want to do that and then merge?

**Comment by danbraunai** [comment]
@chandanms changes look great! Still one comment about the 35M_config.yaml naming. Feel free to merge after handling that.

**Comment by chandanms** [comment]
@danbraunai Made the changes to file. Btw, not sure if I am missing something but I don't think I have the merge PR rights!
